### ISSUE TITLE
Consider decoupled weight decay optimizers?

### ISSUE THREAD

**Issue by abhi-mosaic** [microsoft/mup#1]
Hi! I'm a big fan of this project and I noticed that MUP has some wrappers for common optimizers `MuAdam`, `MuAdamW`, and `MuSGD`. Given the focus on hyperparameter stability and tuning, I was wondering if you might be interested in adding / experimenting with the decoupled weight decay optimizers from this paper (https://arxiv.org/abs/1711.05101)? 

For context, PyTorch's implementations of `SGD`, `Adam`, and even `AdamW` all scale the weight decay by the learning rate in their update step, and I've found that this makes it tough to tune the two values independently (if you increase LR, you also silently increase the effective WD). This is due to PyTorch's scheduler implementation which only updates an Optimizer's LR, and so they schedule the WD in sync by multiplying the two values together. 

E.g. here is Pytorch's AdamW code, which shows why it is not really decoupled: https://github.com/pytorch/pytorch/blob/11231b0f935c092598c994a4bab686485aac1856/torch/optim/adamw.py#L248

The "correct" way to decouple LR, WD is described in the paper, and we have some PyTorch-ready implementations here ([code](https://docs.mosaicml.com/en/v0.4.0/_modules/composer/optim/decoupled_weight_decay.html), [docs](https://docs.mosaicml.com/en/stable/method_cards/decoupled_weight_decay.html)) in MosaicML's Composer library. Though I haven't seen any examples in MUP yet of tuning the WD across model sizes, I feel like this could be a common hparam that users want to tune and that `DecoupledSGDW` or `DecoupledAdamW` could help make it more stable :)

**Comment by thegregyang** [comment]
Hi Abhi,

Thanks for the kind words. For such decoupled optimizers, you can comment out this line https://github.com/microsoft/mup/blob/main/mup/optim.py#L68 for Adam to scale properly, and https://github.com/microsoft/mup/blob/main/mup/optim.py#L113 and https://github.com/microsoft/mup/blob/main/mup/optim.py#L116 for SGD to scale properly.

While I agree with you that this weight decay decoupling makes more sense to me than the way it's done right now, practically speaking I don't have a sense how common it is for people to use it. So I'll have to think about the right interface to expose this kind of flag.

**Comment by abhi-mosaic** [comment]
Hi Greg,

Thanks for the quick response! So if I'm reading this right, were those lines written to basically "undo" the PyTorch scaling?

**Comment by thegregyang** [comment]
Yep you can think of it like that.

**Comment by thegregyang** [comment]
Hey @abhi-mosaic, we have added a flag for decoupling weight decay now. See https://github.com/microsoft/mup/blob/main/mup/optim.py#L47. I'll close the issue now but feel free to reopen if there's any further problems.

**Comment by SeunghyunSEO** [comment]
hi guys, i'm a huge fan of MuP too.
glad to see this issue, but i cant understand what [decoupled_wd](https://github.com/microsoft/mup/blob/main/mup/optim.py#L77-L79) flag do.
when user want to use AdamW, should decoupled_wd be activated?

if it's right, like @abhi-mosaic said,

```
param = (1 - default_lr*lr_schedule_multiplier*weight_decay) * param
```

pytorch default adamw can cause scaling weight_decay factor depending on width because default_lr for this param group is already scaled by `1/width_mult`.
so i think it's fair to set `weight_decay = original_weight_decay*width_mult/lr` if we want to use truly decoupled adamw like @thegregyang did for naive Adam optim class

**Comment by karibbov** [comment]
@SeunghyunSEO 
A late reply, but in case anyone else was confused like me, [decoupled_wd](https://github.com/microsoft/mup/blob/19814971934ef91dd546f88e913fc963e096d11c/mup/optim.py#L47) flag should only be set if the underlying optimizer implementation ([impl](https://github.com/microsoft/mup/blob/19814971934ef91dd546f88e913fc963e096d11c/mup/optim.py#L45) argument here) you're using is decoupled. (i.e. You're setting `impl=`[DecoupledAdamW](https://docs.mosaicml.com/projects/composer/en/latest/api_reference/generated/composer.optim.DecoupledAdamW.html))
